types/tipset.go

```diff
@@ -2,155 +2,148 @@ package types
+	// ErrEmptyTipSet is returned no blocks are provided to a tipset contructor
+	ErrEmptyTipSet = errors.New("no blocks for tipset")
 )
 
-// NewTipSet returns a TipSet wrapping the input blocks.
-// PRECONDITION: all blocks are the same height and have the same parent set.
-func NewTipSet(blks ...*Block) (TipSet, error) {
-	if len(blks) == 0 {
-		return nil, errors.New("Cannot create tipset: no blocks")
-	}
-	ts := TipSet{}
-	for _, b := range blks {
-		if err := ts.AddBlock(b); err != nil {
-			return nil, errors.Wrapf(err, "Cannot create tipset")
+// NoTipSet is a singleton representing a nil or undefined tipset.
+var NoTipSet = TipSet{}
+
+// NewTipSet builds a new TipSet from a collection of blocks.
+// The blocks must be distinct (different CIDs), have the same height, and same parent set.
+func NewTipSet(blocks ...*Block) (TipSet, error) {
+	if len(blocks) == 0 {
+		return NoTipSet, ErrEmptyTipSet
+	}
+
+	first := blocks[0]
+	height := first.Height
+	parents := first.Parents
+	weight := first.ParentWeight
+	cids := make(map[cid.Cid]bool)
+
+	sorted := make([]*Block, len(blocks))
+	for i, blk := range blocks {
+		if i > 0 { // Skip redundant checks for first block
+			if blk.Height != height {
+				return NoTipSet, errors.Errorf("Inconsistent block heights %d and %d", height, blk.Height)
+			}
+			if !blk.Parents.Equals(parents) {
+				return NoTipSet, errors.Errorf("Inconsistent block parents %s and %s", parents.String(), blk.Parents.String())
+			}
+			if blk.ParentWeight != weight {
+				return NoTipSet, errors.Errorf("Inconsistent block parent weights %d and %d", weight, blk.ParentWeight)
+			}
 		}
-	}
-	return ts, nil
-}
-
-// AddBlock adds the provided block to this tipset.
-// PRECONDITION: this block has the same height parent set as other members of ts.
-func (ts TipSet) AddBlock(b *Block) error {
-	if len(ts) == 0 {
-		id := b.Cid()
-		ts[id] = b
-		return nil
+		// Reject duplicate blocks (by CID).
+		c := blk.Cid()
+		if cids[c] {
+			return NoTipSet, errors.Errorf("Duplicate block CID %s", c)
+		}
+		cids[c] = true
+		sorted[i] = blk
 	}
 
-	h, err := ts.Height()
-	if err != nil {
-		return err
-	}
-	p, err := ts.Parents()
-	if err != nil {
-		return err
-	}
-	weight, err := ts.ParentWeight()
-	if err != nil {
-		return err
-	}
-	if uint64(b.Height) != h {
-		return errors.Errorf("block height %d doesn't match existing tipset height %d", uint64(b.Height), h)
-	}
-	if !b.Parents.Equals(p) {
-		return errors.Errorf("block parents %s don't match tipset parents %s", b.Parents.String(), p.String())
-	}
-	if uint64(b.ParentWeight) != weight {
-		return errors.Errorf("bBlock parent weight: %d doesn't match existing tipset parent weight: %d", uint64(b.ParentWeight), weight)
-	}
+	// Sort blocks by ticket.
+	sort.Slice(sorted, func(i, j int) bool {
+		cmp := bytes.Compare(sorted[i].Ticket, sorted[j].Ticket)
+		if cmp == 0 {
+			// Break ticket ties with the block CIDs, which are distinct.
```

Comment:
Lets make sure there is an issue somewhere (gfc or spec) tracking that this is recorded in the spec.

Reply:
I'll do this when fixing the spec generally to define this. https://github.com/filecoin-project/specs/pull/289